### PR TITLE
Add win_uri Module

### PR DESCRIPTION
**Pull request by blakfeld**
An attempt to port the "uri" module to Windows using "Invoke-WebRequest".

### CODE REVIEW

**Comment by robynbergeron** [comment]
Hi @blakfeld --

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

**Comment by robynbergeron** [comment]
Pinging our windows crew also as this may be of interest to them. 
@trondhindenes @petemounce @elventear @smadam813 @jhawkesworth @angstwad @sivel @chrishoffman @cchurch

**Comment by davidobrien1985** [comment]
I like it :) :+1:

File: windows/win_uri.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# this is a windows documentation stub.  actual code lives in the .ps1
+# file of the same name
+
+DOCUMENTATION = """
+---
+module: win_uri
+version_added: ""
```
Comment (jhawkesworth) [inline windows/win_uri.py]: please put in version - current dev is "2.0" at the time of writing.

**Comment by jhawkesworth** [comment]
This would be useful.  
Can you do (very minor) doc fixes please?
--jon

**Comment by blakfeld** [comment]
Documentation fixed! I also poked through the official docs, and added "required" and better "Default" fields.

**Comment by robynbergeron** [comment]
Hi @blakfeld -- thanks for the docs fixes!

@davidobrien1985 & @jhawkesworth -- a super friendly request that we need a +1 vote for each of the following things (as described above... but I'll paste it again for good measure):

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

Thanks for your help -- great to see new modules getting reviews! :dancers:

**Comment by jhawkesworth** [comment]
ok, got back to this again.

Thanks for the doc fixes.

I know I'm being really picky here but obviously want to make it all as good as possible.

Using Get-AnsibleParam might reduce the lines of code a bit - I like small modules as they transfer to hosts more quickly.

Really tiny thing - adding your github id to the Author field in the docs would make you popular with the ansible maintainers too.

If you have any test playbooks can you share to give others a hand to get started testing too?

Jon

**Comment by blakfeld** [comment]
@jhawkesworth -- Thanks for the nit-picks! Nit-picks are the only way to get quality code, and I'm a Linux guy by trade, so my Powershell likely needs nit-picking.

I went and switched over to use Get-AnsibleParams, which seems like a way better way to do what I was doing.

I added my GitHub name to the docs

I don't have a formal test playbook, but here's the task I'm using it for in production:

```
---
- name: Ensure Build is Available on Fileserver
  when: ensure_build
  win_uri:
    url: "http://www.somesite.com"
    method: HEAD
  register: build_check_output
  until: build_check_output.StatusCode == 200
  retries: 30
  delay: 10
```

**Comment by blakfeld** [comment]
I've run into a bug with the updated params. I'm working on it now, I'll update when I get it sorted

EDIT:

Okay, sorted it all out! I just updated the new Params too quickly and didn't proof-read as much as I should've

**Comment by gregdek** [comment]
Back to @jhawkesworth ? :)

**Comment by jhawkesworth** [comment]
Thanks for turning this round so quickly @blakfeld 
Giving this a meets_guidelines +1 now.

**Comment by blakfeld** [comment]
Anyone opposed to merging this?

**Comment by blakfeld** [comment]
Found and fixed a bug. If you ran this with a service account that had never used Internet Explorer, it would fail. Under the covers I'm using Invoke-WebRequest, which uses the Internet Explorer Engine to parse the web page, and if you've never run Internet Explorer, some configuration isn't set, so it fails. The solution is to use the -UseBasicParsing flag. The only upside to using the IEEngine is that it allows the DOM to be used like a Powershell object, but since that isn't super useful to Ansible, I defaulted that to true.

**Comment by jhawkesworth** [comment]
Good catch with the -UseBasicParsing and I like that you have explained it in the docs too.

Since you have revised, I checked again and this still

meets_guidelines

Just needs someone (other than me) to test now.

Jon

**Comment by nitzmahone** [comment]
@blakfeld Looks pretty good- I played around with it a bit and it behaved as expected. One minor request, though- can you insert a little something to translate the return keys into ansible standard-looking return values (eg, as you're adding the keys to the result object, roll them through a dumb func to convert PasalCasedStringKey into pascal_cased_string_key), and adjust the docs accordingly? @bcoca or @cchurch might have a different opinion, but I think we want to abstract the result dict so it looks like a normal Ansible result dict.

**Comment by blakfeld** [comment]
@nitzmahone seems reasonable. Hacked that in!

**Comment by nitzmahone** [comment]
TIL "snake_case"... Sold!

**Pull request merged**